swiftide-integrations/src/pgvector/mod.rs

```diff
@@ -0,0 +1,163 @@
+//! This module integrates with the pgvector database, providing functionalities to create and manage vector collections,
+//! store data, and optimize indexing for efficient searches.
+//!
+//! pgvector is utilized in both the `indexing::Pipeline` and `query::Pipeline` modules.
+mod persist;
+mod pgv_table_types;
+use anyhow::Result;
+use derive_builder::Builder;
+use sqlx::PgPool;
+use std::fmt;
+
+use pgv_table_types::{FieldConfig, MetadataConfig, PgDBConnectionPool, VectorConfig};
+
+const DEFAULT_BATCH_SIZE: usize = 50;
+
+/// Represents a Pgvector client with configuration options.
+///
+/// This struct is used to interact with the Pgvector vector database, providing methods to manage vector collections,
+/// store data, and ensure efficient searches. The client can be cloned with low cost as it shares connections.
+#[derive(Builder, Clone)]
+#[builder(setter(into, strip_option), build_fn(error = "anyhow::Error"))]
+pub struct PgVector {
+    /// Database connection pool.
+    #[builder(default = "PgDBConnectionPool::default()")]
+    connection_pool: PgDBConnectionPool,
+
+    /// Table name to store vectors in.
+    #[builder(default = "String::from(\"swiftide_pgv_store\")")]
+    table_name: String,
+
+    /// Default sizes of vectors. Vectors can also be of different
+    /// sizes by specifying the size in the vector configuration.
+    vector_size: Option<i32>,
+
+    /// Batch size for storing nodes.
+    #[builder(default = "Some(DEFAULT_BATCH_SIZE)")]
+    batch_size: Option<usize>,
```

Comment:
Is it intentional that this value is optional and can be None?

Reply:
For the `batch_size` configuration, I decided to keep it optional, drawing inspiration from the reference Qdrant implementation. By default, the batch size for PGVector is set to `50`—though we still plan to fine-tune this for optimal performance.

To answer your question, *"Is it intentional that this value is optional and can be None?"*—yes, it’s designed that way. In the indexing pipeline, `Pipeline::then_store_with()` functions as an adapter, routing `Node Streams` to backend storage depending on the batch setting.

Here’s the approach:
- When `Persist::batch_size()` returns `None`, each node is processed individually, with `Persist::store()` sending each chunk to the backend.
- If `Persist::batch_size()` returns `Some()`, batch processing is enabled. The stream of nodes is grouped into chunks based on the batch size, and `Persist::batch_store()` sends these batches to storage.

Does this implementation align with requirements? Let me know if there’s anything specific you’d like adjusted.